src/main/java/org/javarosa/xpath/analysis/XPathAccumulatingAnalyzer.java

```diff
@@ -0,0 +1,106 @@
+package org.javarosa.xpath.analysis;
+
+import org.javarosa.core.model.instance.TreeReference;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import io.reactivex.annotations.Nullable;
+
+/**
+ * A type of XPathAnalyzer which collects and aggregates a specified type of information from
+ * wherever it is present in the expression.
+ *
+ * IMPORTANT NOTE: An accumulating analyzer may analyze the same sub-expression or context ref of
+ * an expression multiple times in a single analysis pass. This means:
+ * - An AccumulatingAnalyzer is NOT appropriate to use for answering questions such as
+ * "How many times is X referenced in this expression?"
+ * - An accumulating analyzer IS appropriate to use for answering questions such as
+ * "What is the set of all things of X type which are referenced at least one time in this expression?"
+ *
+ * @author Aliza Stone
+ */
+public abstract class XPathAccumulatingAnalyzer<T> extends XPathAnalyzer {
+
+    protected List<T> accumulatedList;
+
+    public XPathAccumulatingAnalyzer() {
+        super();
+    }
+
+    public XPathAccumulatingAnalyzer(TreeReference contextRef) {
+        super(contextRef);
+    }
+
+    @Nullable
+    public Set<T> accumulate(XPathAnalyzable rootExpression) {
```

Comment:
Seems weird for us to aggregate a list and then put it into a set on response. Was there a specific reason for that?

Reply:
See the class-level Javadoc for context on this. I wanted to emphasize in-code the point which I make there by not having the accumulation return a List, since that would give the impression that duplicate elements have a useful meaning. Since they don't, a Set seemed like the correct output.

Reply:
Hey,

That makes lots of sense, I think my question was more why we were gathering the intermediate results as a list in the first place, rather than starting with a set. One big advantage of a set is that it would let us add elements to process "in-place" rather than aggregating into a set of recursive lists and then aggregating those back down. For big expressions this approach feels like it'll generate a big tree of strings that then compress back down to 2 strings, when it would have worked to just keep track of the set of 2 strings during the process.